### ISSUE TITLE
Add APIs to make it possible for users of RouteEndpointBuilder to use filters

### ISSUE THREAD

**Issue by davidfowl** [dotnet/aspnetcore#42592]
## Background and Motivation

In .NET 7 preview 6 we made it possible to run RouteHandlerFilter (name TBD) on all endpoints. That implementation always wraps the `RequestDelegate` so that all RouteEndpoint(s) will execute filters. We need to make it possible for custom EndpointDataSource to get access to the filter list on the RouteEndpointBuilder.

## Proposed API

```diff
namespace Microsoft.AspNetCore.Routing;

public static class RouteEndpointBuilder 
{
+    public List<Func<RouteHandlerContext, RouteHandlerFilterDelegate, RouteHandlerFilterDelegate>>? RouteHandlerFilterFactories { get; set; }
}
```

## Usage Examples

It's non-trivial to show real usage. Generally these filters need to be composed around a delegate. For that, they need to be exposed.

## Risks

None

**Comment by davidfowl** [comment]
It needs to be mutable (this is a builder API). IList<T> would also be fine.

**Comment by davidfowl** [comment]
@halter73 not sure, saving on an allocation per route seems worth it 😅

**Comment by halter73** [comment]
Based on the conversation [here](https://github.com/dotnet/aspnetcore/pull/42827#discussion_r933699006), we want to both update this name to `FilterFactories` since it affects more than just "route handlers" now and we want to move the property from `RouteEndpointBuilder` to `EndointBuilder` for similar reasons. Here is the new API proposal:


```diff
namespace Microsoft.AspNetCore.Builder;

public abstract class EndpointBuilder 
{
+    public IList<Func<EndpointFilterFactoryContext, EndpointFilterDelegate, EndpointFilterDelegate>> FilterFactories { get; }
}
```

**Comment by halter73** [comment]
API Review Notes:

- Yes. This does force an extra list allocation for every `EndpointBuilder` instance.
  - We do not care too much about this considering this is already the case for the `Metadata` list and is consistent.
- We seem generally happy with the shorter name and moving the property to the base class.

```diff
namespace Microsoft.AspNetCore.Builder;

public abstract class EndpointBuilder 
{
+    public IList<Func<EndpointFilterFactoryContext, EndpointFilterDelegate, EndpointFilterDelegate>> FilterFactories { get; }
}
```

**Comment by davidfowl** [comment]
> We do not care too much about this considering this is already the case for the Metadata list and is consistent.

😢

**Comment by halter73** [comment]
Fixed by #42557.

**Comment by davidfowl** [comment]
Thanks!
